### DISCUSSION TITLE
where is the clang-format-17 on Ubuntu 22.04 coming from and what is its exact version number?

### DISCUSSION THREAD

**Discussion by jeffli678** [jidicula/clang-format-action#179] [Q&A]
I have a project that uses this action to check clang-format. I am currently using clang-format 17 to locally format the code. However, the formatted source code results in error in the action run. I am guessing this is caused by subtle differences in the clang-format version. 

FYI, I am using the clang-format-17 from msys2 and it is 17.0.6. What is the version of clang-format-17 this action is using? I checked the ones that come with ubuntu 22.04 source, and it does not have clang-format-17. It only has clang-format 13/14/15. So I am wondering where is the clang-format-17 used by this action coming from, and what is its exact number?

**Comment by jeffli678** [comment]
oh nvm I figured it is `Ubuntu clang-format version 17.0.2 (1~exp1ubuntu2.1)`. This could be the problem -- since I am using 17.0.6

**Reply by jeffli678** [reply]
I guess this can be fixed by updating the docker to use the latest sub-version of clang-format-17?

**Comment by jidicula** [comment]
Here's a quick walkthrough of what version of `clang-format` I install. In the `clang-format-image.yml` workflow, I've defined a [mapping of each `clang-format` version to an Ubuntu release](https://github.com/jidicula/clang-format-action/blob/dbc2305aa6bf6cad6e43dbe9679f8ae2094342c3/.github/workflows/clang-format-image.yml#L30-L45) that contains it:

https://github.com/jidicula/clang-format-action/blob/dbc2305aa6bf6cad6e43dbe9679f8ae2094342c3/.github/workflows/clang-format-image.yml#L30-L45

I've grabbed these mappings after looking at package.ubuntu.com and [searching for `clang-format` across all suites](https://packages.ubuntu.com/search?suite=all&arch=any&searchon=names&keywords=clang-format).

This Actions workflow runs daily to rebuild and republish a Docker image defined in [this Dockerfile](https://github.com/jidicula/clang-format-action/blob/dbc2305aa6bf6cad6e43dbe9679f8ae2094342c3/clang-format-docker/Dockerfile). This Dockerfile accepts 2 args:
* `UBUNTU_VERSION`
* `CLANG_FORMAT_VERSION`

These 2 args are used to first fetch the base Docker image for the Ubuntu version I want to build from, then to `apt install` the particular `clang-format` version I want to include in the Docker image. This will always get the latest minor or patch version of `clang-format` available in Ubuntu's apt archives at image build time. You can see the [final images here](https://github.com/jidicula/clang-format-action/pkgs/container/clang-format).

---

This makes it a bit tricky for me to include particular minor or patch versions of `clang-format` in this Action. Versions 3.9-6.0 are sort of an exception because I _think_ the LLVM project was doing things differently then, but in any case I've selected only 1 minor version under each major version to support.

I know this probably doesn't really answer your question satisfactorily - I made this tradeoff to only provide guarantees at the major version level so I'd be able to keep a consistent build process across all versions (3.9 to 18) just with a simple `apt install`. I did look into the alternative of doing a proper build myself, but this seemed to lead down the path of building up the whole LLVM toolchain and build dependencies in an Actions workflow and building the binaries there, which was very challening. Bootstrapping old versions of `cmake` & co. didn't seem simple. I'd definitely welcome a PR if anyone is able to wrangle a build process that's consistent across all versions and is relatively easy to maintain.

**Reply by echoix** [reply]
Did you ever stumble upon https://github.com/muttleyxd/clang-tools-static-binaries?

There's also this that builds clang-format, but is targeted at building python wheels https://github.com/ssciwr/clang-format-wheel

**Reply by jidicula** [reply]
That's interesting, I didn't know about these. Pulling static binaries is interesting, but given recent news about what happened with `xz-utils`, I think I won't adopt either of these as a dependency, considering the wide footprint of my Action's users. The build provenance in those repos is nice, but I wouldn't be aware of changes in those projects' maintainership unless I'm watching those repos.
